estimator/ntru_primal.py

```diff
@@ -0,0 +1,503 @@
+
+            norm_threshold = exp(2*(B_shape[s-beta]))/sigma_sq
+            proba_one = chisquared_table[beta].cum_distribution_function(norm_threshold)
+
+            if proba_one <= 10e-8:
+                continue
+
+            # account for pulling back probability if beta small
+            if beta <= 20:
+                for j in range(2, int(s/beta+1)):
+                    if proba_one < 10**(-6):
+                        proba_one = 0.
+                        break
+                    ind = s - j*(beta-1)-1
+                    norm_bt = exp(2*B_shape[ind])/sigma_sq
+                    norm_b2 = exp(2*B_shape[ind+beta-1])/sigma_sq
+                    proba_one *= conditional_chi_squared(beta-1, s-ind-(beta-1), norm_bt, norm_b2)
+
+            prob_pos[s-beta] = proba_one
+            prob_all_not *= max(1.-proba_one, 0.)
+            Logging.log("dsd", log_level+1, f"Pr[dsd, {beta}] = {prob_all_not}")
+
+        return RR(1.-prob_all_not), prob_pos
+
+    def __call__(
+        self,
+        params: NTRUParameters,
+        red_cost_model=red_cost_model_default,
+        red_shape_model=red_shape_model_default,
+        log_level=1,
+        **kwds,
+    ):
+        """
+        Estimate cost of solving (overstretched) NTRU using the Dense sublattice.
+        Code is adapted from Léo Ducas and Wessel van Woerden.
+        See https://github.com/WvanWoerden/NTRUFatigue for original source
+
+        :param params: NTRU parameters.
+        :param red_cost_model: How to cost lattice reduction.
+        :param red_shape_model: How to model the shape of a reduced basis.
+        :return: A cost dictionary.
+
+        The returned cost dictionary has the following entries:
+
+        - ``rop``: Total number of word operations (≈ CPU cycles).
+        - ``red``: Number of word operations in lattice reduction.
+        - ``δ``: Root-Hermite factor targeted by lattice reduction.
+        - ``β``: BKZ block size.
+        - ``d``: Lattice dimension.
+
+        EXAMPLE::
+
+            >>> from estimator import *
+            >>> NTRU.primal_dsd(schemes.NTRUHPS2048509Enc)
+            rop: ≈2^157.1, red: ≈2^157.1, δ: 1.003645, β: 453, d: 1016, tag: dsd
+
+            NOTE: Non-overstretched parameters (where the probability of Dense sublattice
+            discovery is 0) will return β = d.
+
+            >>> params = NTRU.Parameters(n=113, q=512, Xs=ND.UniformMod(3), Xe=ND.UniformMod(3))
+            >>> NTRU.primal_dsd(params, red_shape_model="zgsa")
+            rop: ≈2^41.3, red: ≈2^41.3, δ: 1.012468, β: 42, d: 226, tag: dsd
+
+            >>> NTRU.primal_dsd(params, red_shape_model="cn11")
+            rop: ≈2^41.2, red: ≈2^41.2, δ: 1.012468, β: 41, d: 226, tag: dsd
+
+            >>> NTRU.primal_dsd(params, red_shape_model=Simulator.CN11)
+            rop: ≈2^41.2, red: ≈2^41.2, δ: 1.012468, β: 41, d: 226, tag: dsd
+
+        The success condition was formulated in [EC:KF17] and further refined in [AC:DvW21]
```

Comment:
Add those to references.rst, also note that for two authors we give three letters per name (same as cryptobib)

Reply:
Done as of latest commit.